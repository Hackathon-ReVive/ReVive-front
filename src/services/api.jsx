import axios from "axios";

const API_BASE_URL = "http://localhost:8080/api";

export const loginUser = async (userData) => {
  try {
    const response = await axios.post(`${API_BASE_URL}/login`, userData);
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message;
  }
};

export const registerUser = async (userData) => {
  try {
    const response = await axios.post(`${API_BASE_URL}/register`, userData);
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message;
  }
};

export const getUserById = async (id) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/users/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error al buscar usuario:", error);
    throw error;
  }
};

export const updateUser = async (id, userData) => {
  try {
    const response = await axios.put(`${API_BASE_URL}/users/${id}`, userData, {
      headers: {
        "Content-Type": "application/json",
      },
    });
    return response.data;
  } catch (error) {
    console.error("Error al actualizar usuario:", error);
    throw error;
  }
};

export const deleteUser = async (id) => {
  try {
    const response = await axios.delete(`${API_BASE_URL}/users/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error al borrar usuario", error);
    throw error;
  }
};

// Productos
export const getProducts = async () => {
  try {
    const response = await axios.get(`${API_BASE_URL}/products`);
    return response.data;
  } catch (error) {
    console.error("Error al buscar productos:", error);
    throw error;
  }
};

export const getProductById = async (id) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/products/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error al buscar producto:", error);
    throw error;
  }
};

export const createProduct = async (productData) => {
  try {
    const response = await axios.post(`${API_BASE_URL}/products`, productData, {
      headers: {
        "Content-Type": "application/json",
      },
    });
    return response.data;
  } catch (error) {
    console.error("Error al crear producto:", error);
    throw error;
  }
};

export const updateProduct = async (id, productData) => {
  try {
    const response = await axios.put(
      `${API_BASE_URL}/products/${id}`,
      productData,
      {
        headers: {
          "Content-Type": "application/json",
        },
      }
    );
    return response.data;
  } catch (error) {
    console.error("Error al actualizar producto:", error);
    throw error;
  }
};

export const deleteProduct = async (id) => {
  try {
    const response = await axios.delete(`${API_BASE_URL}/products/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error al borrar producto:", error);
    throw error;
  }
};

// Órdenes
export const getOrders = async () => {
  try {
    const response = await axios.get(`${API_BASE_URL}/orders`);
    return response.data;
  } catch (error) {
    console.error("Error al buscar pedidos:", error);
    throw error;
  }
};

export const getOrderById = async (id) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/orders/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error al buscar pedido:", error);
    throw error;
  }
};

export const createOrder = async (orderData) => {
  try {
    const response = await axios.post(`${API_BASE_URL}/orders`, orderData, {
      headers: {
        "Content-Type": "application/json",
      },
    });
    return response.data;
  } catch (error) {
    console.error("Error al crear pedido:", error);
    throw error;
  }
};

export const updateOrder = async (id, orderData) => {
  try {
    const response = await axios.put(
      `${API_BASE_URL}/orders/${id}`,
      orderData,
      {
        headers: {
          "Content-Type": "application/json",
        },
      }
    );
    return response.data;
  } catch (error) {
    console.error("Error al actualizar pedido:", error);
    throw error;
  }
};

export const deleteOrder = async (id) => {
  try {
    const response = await axios.delete(`${API_BASE_URL}/orders/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error al borrar pedido:", error);
    throw error;
  }
};

// Carrito
export const getCart = async (userId) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/users/${userId}/cart`);
    return response.data;
  } catch (error) {
    console.error("Error al buscar el carro:", error);
    throw error;
  }
};

export const addToCart = async (userId, product) => {
  try {
    const response = await axios.post(
      `${API_BASE_URL}/users/${userId}/cart`,
      product,
      {
        headers: {
          "Content-Type": "application/json",
        },
      }
    );
    return response.data;
  } catch (error) {
    console.error("Error al añadir al carro:", error);
    throw error;
  }
};

export const removeFromCart = async (userId, productId) => {
  try {
    const response = await axios.delete(
      `${API_BASE_URL}/users/${userId}/cart/${productId}`
    );
    return response.data;
  } catch (error) {
    console.error("Error al revmover del carro:", error);
    throw error;
  }
};
