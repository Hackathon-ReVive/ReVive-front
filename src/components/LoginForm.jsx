import React from "react";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { loginUser } from "../services/api";
import Button from "./Button";

function LoginForm() {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();
  const navigate = useNavigate();

  const onSubmit = async (data) => {
    try {
      const userData = await loginUser(data);
      console.log("Login Exitoso:", userData);
      navigate("/");
    } catch (error) {
      console.error("Error en el Login:", error);
    }
  };

  return (
    <section className="min-h-screen flex items-center justify-center bg-cover bg-center bg-[url(/assets/bgLoginRegister.png)]">
      <div className="max-w-md mx-auto p-8 bg-[#ECFDF5] shadow-md rounded-3xl border border-green-300 backdrop-blur-lg bg-opacity-80">
        <h2 className="text-3xl font-bold text-center text-[#064E3B] mb-2">
          Conéctate
        </h2>
        <p className="text-center text-[#059669] mb-4">
          ¡Dale vida nueva a tus productos!
        </p>

        <form onSubmit={handleSubmit(onSubmit)} className="space-y-4">
          <label
            htmlFor="email"
            className="block text-[#059669] text-sm font-medium"
          >
            Correo Electrónico
          </label>
          <input
            type="email"
            {...register("email", {
              required: "Email es obligatorio",
              pattern: {
                value: /^[^@\s]+@[^@\s]+\.[a-zA-Z]{2,}$/i,
                message: "Formato de email inválido",
              },
            })}
            placeholder="Email Address"
            className="w-full p-3 border border-[#A7F3D0] rounded-lg"
            id="email"
          />
          {errors.email && (
            <p className="text-red-500 text-sm border-1 border-red-500 rounded bg-red-100 p-2">
              {errors.email.message}
            </p>
          )}


        <form onSubmit={handleSubmit(onSubmit)} className="space-y-4">
          <label
            htmlFor="email"
            className="block text-[#059669] text-sm font-medium"
          >
            Correo Electrónico
          </label>
          <input
            type="email"
            {...register("email", {
              required: "Email es obligatorio",
              pattern: {
                value: /^[^@\s]+@[^@\s]+\.[a-zA-Z]{2,}$/i,
                message: "Formato de email inválido",
              },
            })}
            placeholder="Email Address"
            className="w-full p-3 border border-[#A7F3D0] rounded-lg"
            id="email"
          />
          {errors.email && (
            <p className="text-red-500 text-sm border-1 border-red-500 rounded bg-red-100 p-2">
              {errors.email.message}
            </p>
          )}

          <label
            htmlFor="password"
            className="block text-[#059669] text-sm font-medium"
          >
            Contraseña
          </label>
          <input
            type="password"
            {...register("password", { required: "Contraseña es obligatoria" })}
            placeholder="Password"
            className="w-full p-3 border border-[#A7F3D0] rounded-lg bg-[#ECFDF5]"
            id="password"
          />
          {errors.password && (
            <p className="text-red-500 text-sm border-1 border-red-500 rounded bg-red-100 p-2">
              {errors.password.message}
            </p>
          )}

          <Button
            type="submit"
            text="Continuar con Email"
            className="w-full text-white rounded-lg transition"
          />
        </form>
        <Button
          text="No tengo cuenta"
          className="mt-4 border-[#A7F3D0] border w-full rounded-lg"
          bgColor="bg-[#DCFCE7]"
          txtColor="text-teal-600"
          onClick={() => navigate("/register")}
        />
      </div>
    </section>
  );
}

export default LoginForm;
