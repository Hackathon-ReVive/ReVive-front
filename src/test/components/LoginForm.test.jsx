import "@testing-library/jest-dom";
import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import LoginForm from "../../components/LoginForm";
import { loginUser } from "../../services/api";

jest.mock("../../services/api", () => ({
  loginUser: jest.fn(),
}));

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("LoginForm Component", () => {
  
  test("shows validation errors when submitting empty fields", async () => {
    render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByRole("button", { name: /Continuar con Email/i }));

    expect(await screen.findByText(/Email es obligatorio/i)).toBeInTheDocument();
    expect(await screen.findByText(/ContraseÃ±a es obligatoria/i)).toBeInTheDocument();
  });

  
  test("calls loginUser when form is submitted with valid data", async () => {
    loginUser.mockResolvedValueOnce({ token: "mockToken" });

    render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );

    fireEvent.change(screen.getByPlaceholderText("Email Address"), { target: { value: "test@example.com" } });
    fireEvent.change(screen.getByPlaceholderText("Password"), { target: { value: "password123" } });
    fireEvent.click(screen.getByRole("button", { name: /Continuar con Email/i }));

    await waitFor(() => {
      expect(loginUser).toHaveBeenCalledWith({
        email: "test@example.com",
        password: "password123",
      });
    });
  });

  test("navigates to register page when 'No tengo cuenta' button is clicked", () => {
    render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );
  
    fireEvent.click(screen.getByRole("button", { name: /No tengo cuenta/i }));
  
    expect(mockNavigate).toHaveBeenCalledWith("/register");
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

});

