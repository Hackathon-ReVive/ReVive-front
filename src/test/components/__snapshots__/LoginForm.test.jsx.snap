// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="min-h-screen flex items-center justify-center bg-cover bg-center bg-[url(/assets/bgLoginRegister.png)]"
  >
    <div
      class="max-w-md mx-auto p-8 bg-[#ECFDF5] shadow-md rounded-3xl border border-green-300 backdrop-blur-lg bg-opacity-80"
    >
      <h2
        class="text-3xl font-bold text-center text-[#064E3B] mb-2"
      >
        Conéctate
      </h2>
      <p
        class="text-center text-[#059669] mb-4"
      >
        ¡Dale vida nueva a tus productos!
      </p>
      <form
        class="space-y-4"
      >
        <label
          class="block text-[#059669] text-sm font-medium"
          for="email"
        >
          Correo Electrónico
        </label>
        <input
          class="w-full p-3 border border-[#A7F3D0] rounded-lg"
          id="email"
          name="email"
          placeholder="Email Address"
          type="email"
        />
        <label
          class="block text-[#059669] text-sm font-medium"
          for="password"
        >
          Contraseña
        </label>
        <input
          class="w-full p-3 border border-[#A7F3D0] rounded-lg bg-[#ECFDF5]"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
        />
        <button
          class="bg-teal-500 text-white p-2 rounded hover:bg-teal-600 hover:text-white cursor-pointer transition w-full text-white rounded-lg transition"
          type="submit"
        >
          Continuar con Email
        </button>
      </form>
      <button
        class="bg-[#DCFCE7] text-teal-600 p-2 rounded hover:bg-teal-600 hover:text-white cursor-pointer transition mt-4 border-[#A7F3D0] border w-full rounded-lg"
        type="button"
      >
        No tengo cuenta
      </button>
    </div>
  </section>
</DocumentFragment>
`;
